L1Trigger/L1TMuon/plugins/L1TDisplacedMuonProducer.cc

```diff
@@ -0,0 +1,268 @@
+// system include files
+#include <memory>
+#include <fstream>
+#include <sstream>
+
+// user include files
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+// geometry
+#include "Geometry/Records/interface/MuonGeometryRecord.h"
+#include "Geometry/GEMGeometry/interface/GEMGeometry.h"
+#include "Geometry/RPCGeometry/interface/RPCGeometry.h"
+#include "Geometry/DTGeometry/interface/DTGeometry.h"
+#include "Geometry/CSCGeometry/interface/CSCGeometry.h"
+#include "Geometry/GEMGeometry/interface/ME0Geometry.h"
+
+#include "L1Trigger/L1TMuon/src/Phase2/L1TDisplacedMuonBuilder.h"
+//
+// class declaration
+//
+using namespace l1t;
+
+class L1TDisplacedMuonProducer : public edm::EDProducer
+{
+public:
+  explicit L1TDisplacedMuonProducer(const edm::ParameterSet&);
+  ~L1TDisplacedMuonProducer();
+
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  virtual void beginJob() ;
+  virtual void produce(edm::Event&, const edm::EventSetup&);
+  virtual void endJob() ;
+
+  virtual void beginRun(edm::Run const&, edm::EventSetup const&);
+  virtual void endRun(edm::Run const&, edm::EventSetup const&);
+  virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&);
+  virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&);
+
+  // ----------member data ---------------------------
+  edm::InputTag cscCompTag_;
+  edm::InputTag cscLctTag_;
+  edm::InputTag padTag_;
+  edm::InputTag copadTag_;
+  edm::InputTag me0TriggerTag_;
+  edm::InputTag me0SegmentTag_;
+  edm::InputTag emtfTag_;
+  edm::InputTag muonTag_;
+  edm::InputTag bmtfTag_;
+  edm::InputTag omtfNegTag_;
+  edm::InputTag omtfPosTag_;
+  edm::InputTag emtfNegTag_;
+  edm::InputTag emtfPosTag_;
+
+  edm::EDGetTokenT<CSCComparatorDigiCollection> comparatorToken_;
+  edm::EDGetTokenT<CSCCorrelatedLCTDigiCollection> lctToken_;
+  edm::EDGetTokenT<GEMPadDigiCollection> padToken_;
+  edm::EDGetTokenT<GEMCoPadDigiCollection> copadToken_;
+  edm::EDGetTokenT<ME0TriggerDigiCollection> me0triggerToken_;
+  edm::EDGetTokenT<ME0SegmentCollection> segmentToken_;
+  edm::EDGetTokenT<l1t::EMTFTrackCollection> emtfToken_;
+  edm::EDGetTokenT<l1t::MuonBxCollection> muonToken_;
+  edm::EDGetTokenT<l1t::MuonBxCollection> bmtfToken_;
```

Comment:
Do you really want the TF collections to be MuonBxCollections and not RegionalMuonCandBxCollections?

Reply:
I'm still working on this. I want to associated the L1Muon to its trigger primitives (CSC, DT, GEM). Unfortunately, the L1Mu does not have accessors. So I was thinking to match the L1Mu to the BMTF/EMTF/OMTF track (with a dR method) and get the stubs that way.

Reply:
but perhaps instead of a dR match I could use a TF Id number...

Reply:
The TF muon index could be used to find out from which TF wedge/sector the track comes but you still have up to three tracks per wedge/sector. A matching is probably needed. In the barrel you could use the track address for the matching. That should be unique. For the others I guess you need a dR matching. The new EMTF emulator should have a track address as well though.

Reply:
@abrinke1 Should I use https://github.com/cms-sw/cmssw/blob/master/DataFormats/L1TMuon/interface/EMTFTrack.h#L150 to match EMTF to L1Mu?

Reply:
Hi Sven,
No, "track_num" is an internal EMTF counter.  In general, I would not use the EMTFTrack collection at all, but rather the RegionalMuonCandBxCollections, which are in the same format for all track finders.  I think.  Maybe you could describe more exactly what you're trying to accomplish, or what the workflow is?  If you're just trying to access current (Phase I) L1T muons to compare to your Phase II studies, I would only use uGMT muons (either input or output), not the private EMTF/OMTF/BMTF collections.  To me, a dR matching must cover 99% of the cases for Phase I vs. Phase II studies.  But maybe I'm missing some complexity, or you're trying to do something else.
Best,
Andrew

Reply:
@abrinke1 I would like to access the CSC, GEM and DT stubs through the track-finder tracks. The stub positions and directions are used directly to calculate the displaced muon pT

Reply:
Ah.  In that case, for EMTF, the EMTFHit within the EMTFTrack (or just the EMTFHitCollection in general) is certainly the simplest solution.  The track stubs within the RegionalMuonCand only contain BX and chamber ID information, not strip/wire ... so they can only be used to match the stubs in the track with EMTFHits (or DTCorrelatedThetaPhiWhatevers) anyway.  If you have to do that, you might as well go straight to the internal track-finder information, however you can get a hold of it.

Reply:
@abrinke1 thanks. sounds good